web/src/compose_validate.ts

```diff
@@ -795,3 +798,82 @@ export function convert_mentions_to_silent_in_direct_messages(
+    return split_messages_enabled;
+}
+
+export function set_split_messages_enabled(enable: boolean): void {
+    split_messages_enabled = enable;
+}
+
+/**
+ *
+ * @param content
+ * @returns index of the delimiter occurring outside code blocks since no splitting should
+ * be done inside code blocks
+ */
+function delimiter_index_outside_code_block(content: string): number {
+    let index = content.indexOf(SPLIT_DELIMITER);
+    // skip delimiters which occur inside code blocks
+    while (index !== -1) {
+        if (!compose_ui.position_inside_code_block(content, index + 1)) {
+            return index;
+        }
+        // find next occurrence of the delimiter
+        index = content.indexOf(SPLIT_DELIMITER, index + 1);
+    }
+    return -1;
+}
+
+/**
+ * Splits a message into 2 parts, if possible
+ * @param message_content the content to be split
+ * @returns a pair of strings: the first being the individual split message, the second being the
+ * remaining part of the message, which will be empty when there are no further parts to send
+ */
+export function split_message(message_content: string): [string, string] {
+    if (!is_split_messages_enabled()) {
+        return [message_content, ""];
+    }
+    const index = delimiter_index_outside_code_block(message_content); // first occurrence of a valid delimiter
+    if (index === -1) {
+        return [message_content, ""];
+    }
+    // if more than 2 blank lines are left, we need to find a non-empty message part to be sent.
+    // so we first use the index to find the remaining content. We use lastIndexOf since it is possible
+    // that rest_content occurs as a substring more than once in message_content if searched from the
+    // beginning
+    const rest_content = message_content.slice(index).trim();
+    const send_content = message_content.slice(0, message_content.lastIndexOf(rest_content));
+    return [send_content, rest_content];
+}
+
+/**
+ * @returns the number of parts the compose content will split into
+ *
+ * mimics the message sending behaviour to count the number of message parts. We cannot directly use
+ * `message_content.split().length` because delimiters inside code blocks should be ignored.
+ */
+export function count_message_content_split_parts(): number {
+    const message_content = compose_state.message_content();
+    let count = 1;
+    let parts = split_message(message_content);
+    while (parts[1]) {
+        count += 1;
+        parts = split_message(parts[1]);
+    }
+    return count;
+}
+
+/**
+ * checks whether the message actually needs to be split, as it is possible
+ * that `split_messages` is enabled but the message doesn't contain the delimiter.
+ */
+export function will_split_into_multiple_messages(): boolean {
+    if (!is_split_messages_enabled()) {
+        return false;
+    }
+    // if there is a non-empty remaining part, splitting will occur
+    return Boolean(split_message(compose_state.message_content())[1]);
```

Comment:
Is  `Boolean()` really needed here?

Reply:
Yes, otherwise it says `Type 'string' is not assignable to type 'boolean'`.